lib/Listerine/channels.ex

```diff
@@ -176,63 +176,57 @@ defmodule Listerine.Channels do
     |> prepend.(remove_course_channels(others))
   end
 
-  def add_role(_message, []) do
-    {:ok, []}
-  end
+  @doc """
+  Adds the roles passed in the `courses` list the the author of the `message`.
 
-  def add_role(message, [name | tail]) do
+  Returns a list of added roles
+  """
+  def add_role(message, courses) do
     guild = message.channel.guild_id
     member = Guild.get_member(guild, message.author.id)
-    roles = get_courses()
-
-    roles =
-      Map.merge(roles["1"], roles["2"])
-      |> Map.merge(roles["3"])
-
-    modified =
-      if roles[name] != nil do
-        case Member.add_role(member, roles[name]["role"]) do
-          :ok ->
-            [name | elem(add_role(message, tail), 1)]
-
-          _ ->
-            elem(add_role(message, tail), 1)
-        end
-      else
-        Message.reply(message, "Role #{name} does not exist")
-        elem(add_role(message, tail), 1)
-      end
 
-    {:ok, modified}
-  end
+    case get_courses() do
+      nil ->
+        nil
+
+      roles ->
+        roles = Enum.reduce(Map.keys(roles), %{}, fn x, acc -> Map.merge(acc, roles[x]) end)
 
-  def rm_role(_message, []) do
-    {:ok, []}
+        Enum.filter(courses, fn x -> x in Map.keys(roles) end)
+        |> Enum.map(fn x ->
+          case Member.add_role(member, roles[x]["role"]) do
+            :ok -> x
+            _ -> "delet_this"
+          end
+        end)
+        |> Enum.filter(fn x -> x != "delet_this" end)
```

Comment:
Can't all this be performed with a reduce? 
Not even thinking about performance (N vs 2N, but it's small) just seems forced with the "delet_this".

Reply:
Done